### ISSUE TITLE
[FEATURE] Storage Tiering support of the Harvester CSI Driver

### ISSUE THREAD

**Issue by dmytrobabeshko** [harvester/harvester#3052]
Hello,

Storage tiering was introduced in Harveser v1.1.0 recently https://github.com/harvester/harvester/issues/2147

Currently, the `harveste-csi-driver` doesn't have the option to have several storage classes which are bound to different storage classes in the harvester.

Currently, the name of the storage class used by the driver is taken from env variable from `harvester-csi-driver-pod`.
https://github.com/harvester/harvester-csi-driver/blob/0f6455e0bc7a4e53dd01ab811eab37182ba6030e/main.go#L39

Thank you in advance!

**Comment by guangbochen** [comment]
this is a known limitation of the v1.1.0 version, we will consider adding support in its future releases, and transfer the issue to the harvester repo.

**Comment by Vicente-Cheng** [comment]
Update the user story and controller flow:

Users want to create volumes with different purposes (which means different performance/capability requirements) on the `harvester` cluster at this guest cluster.

With v1.1.0, we could create different tiers (StorageClass) on the harvester. Now, `harvester-CSI-driver` should support creating PVC on the specific tier on the host.
For example, the user needs 2 replicas for some data, which only needs lower data reliability.
We could create a StorageClass for it like following:
```
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: harvester-replica-2
provisioner: driver.harvesterhci.io
reclaimPolicy: Delete
volumeBindingMode: Immediate
parameters:
  hostTierName: replica-2
```
Then, the user could use this StorageClass `harvester-replica-2` to create PVC with only 2 replica.
If the user wants to create the default tier on the host, use the default StorageClass `harvester` or simply not specify any for it.

The controller flow:

PVC StorageClass would be set up from StorageClass parameter `hostTierName` or controller config `host-storage-class`.
The prioritize is SC parameter > controller config.
If w/o any definition, PVC would use the default StorageClass on the host cluster.

cc @bk201, @guangbochen

**Comment by bk201** [comment]
@guangbochen @yasker Does this sound good to you? 
- User creates a new storage class with `paramters.hostTierName` set to an existing storage class in the host cluster.
- User then creates volumes from the new storage class.

**Comment by yasker** [comment]
Sounds good to me. Also need Guangbo's input. 

However we probably need another name for the field. We didn't mention anything related to the `tier` in the UI IIRC. How about `HostStorageClass` or `HarvesterClusterStorageClass`, or some other names?

**Comment by guangbochen** [comment]
Hi @Vicente Cheng and @Kiefer Chang, I think we can have HEP/design discussion of it if possible, but below are just some questions that I have right now, thanks.
- who is responsible for creating the new type of SC manually on the guest k8s cluster, and how would they know the corresponding host-storage-class name, we may need a design mechanism for this. Also seems RBAC is not applicable here since it is a cluster-wide CRD.
- per new design, we will not define the default SC in the controller, but should we still keep/update the default SC generated on the guest k8s cluster like https://github.com/harvester/charts/blob/master/charts/harvester-csi-driver/templates/storageclass.yaml?
- prefer to keep the name consistent e.g., like `host-storage-class`, also may require some UI effort to make it required when users are trying to create a new SC for the Harvester CSI driver.

**Comment by Vicente-Cheng** [comment]
> Hi @vicente Cheng and @kiefer Chang, I think we can have HEP/design discussion of it if possible, but below are just some questions that I have right now, thanks.
> 
> * who is responsible for creating the new type of SC manually on the guest k8s cluster, and how would they know the corresponding host-storage-class name, we may need a design mechanism for this. Also seems RBAC is not applicable here since it is a cluster-wide CRD.

Maybe the admin user needs to handle it. For now, the user on the guest cluster could not get any StorageClass information from the host cluster because of permission.
Yap, this is the cluster-wide CRD. So RBAC is not working here.
But the behavior currently is the same? I mean that RBAC could not work well on the creating volume in the guest cluster. 

> * per new design, we will not define the default SC in the controller, but should we still keep/update the default SC generated on the guest k8s cluster like https://github.com/harvester/charts/blob/master/charts/harvester-csi-driver/templates/storageclass.yaml?

We still need it. Remove the default SC (I thought you meant removing the hardcode `longhorn`?) because we could let user create volume with default SC on the host harvester cluster. We still need the default SC on the guest cluster for it. 

> * prefer to keep the name consistent e.g., like `host-storage-class`, also may require some UI effort to make it required when users are trying to create a new SC for the Harvester CSI driver.

Agree, I would update the parameter name of StorageClass
For the UI part, we need to let the user could list all SC on the host harvester cluster. This behavior is blocked currently.

---
Hi @guangbochen
I update the HEP on #3318.
And the question reply, please see the above.

Thanks!

**Comment by guangbochen** [comment]
> For the UI part, we need to let the user could list all SC on the host harvester cluster. This behavior is blocked currently.

Users from the guest clusters would not be able to list SCs on the host Harvester cluster, so they may only request it manually through the Harvester cluster admin.

**Comment by Vicente-Cheng** [comment]
The RKE2 code freeze date is Friday, January 13. Just a friendly update from yesterday's meeting.

**Comment by irishgordo** [comment]
@n313893254 
For validating that the banner shows displaying under StorageClasses on Guest Cluster RKE2:
> Please upgrade your Harvester CSI Version to use this feature

![Screenshot from 2023-02-08 15-11-18](https://user-images.githubusercontent.com/5370752/217673970-a1769a4b-388f-47c3-a4f1-56e6b9c0db13.png)
![Screenshot from 2023-02-08 14-46-49](https://user-images.githubusercontent.com/5370752/217673996-87b1dfbb-359c-464e-9072-fec733b4624d.png)

I'm not seeing that on:
- Harvester Version: master-8c620b7d-head 
- Rancher v2.7-head (  0504ce6 )
- Harvester CSI Driver: 0.1.9
- Guest Cluster RKE2: v1.25.6+rke2r1 

After updating RKE-Metadata Config to:
```json
{
  "refresh-interval-minutes": "10",
  "url": "https://harvester-dev.oss-cn-hangzhou.aliyuncs.com/rke-meta-data.json"
}
```
And refreshing metadata.

Did I need to be at an older version than CSI Driver 0.1.9?

**Comment by n313893254** [comment]
Hi, @irishgordo. Sorry for my delayed answer. RKE2 v1.25.6+rke2r1 corresponds to the harvester-csi-driver chart version 0.1.15. It is as expected.

You can use RKE2 v1.24.9+rke2r2 which the version of harvester-csi-driver chart is 0.1.11 to test the banner case. 

[https://releases.rancher.com/kontainer-driver-metadata/dev-v2.7/data.json]
<img width="317" alt="image" src="https://user-images.githubusercontent.com/18737885/218253778-ddaa9d10-ba0b-4c5e-913f-3c354b6d99d0.png">

**Comment by irishgordo** [comment]
@n313893254 thanks for the response back :smile: 

I was wondering, should I be only testing this with Rancher [v2.6-head](https://hub.docker.com/layers/rancher/rancher/v2.6-head/images/sha256-05777b74bb023169d7f98f32eba2be60eec621a5a0d10cd082d608b432c871da?context=explore)?

I'm noticing with:
- Harvester  Version: master-6fd9a5bb-head  ( 02/10/23 Release )
- Rancher v2.7-head
- RKE2 v1.24.9+rke2r2
- CSI 0.1.11

That when I navigate to create a new storage class based on the Harvester CSI, I'm not seeing the banner that warns that the CSI Driver should be upgraded to v0.1.15.

![Screenshot from 2023-02-13 17-12-18](https://user-images.githubusercontent.com/5370752/218612859-e111034f-c69d-46f4-a60c-ae85bd4a0ce4.png)

**Comment by n313893254** [comment]
@irishgordo Thanks for your test. That is a bug. and we should also test with v2.7-head.

**Comment by irishgordo** [comment]
@n313893254 - also noticing the similar behavior with the same loadout except instead of Rancher v2.7-head - noticing it with Rancher v2.6-head:
![Screenshot from 2023-02-14 10-20-30](https://user-images.githubusercontent.com/5370752/218824144-e24a0e73-58d3-4ad9-86df-008645655a9b.png)

**Comment by irishgordo** [comment]
**validation failed:**

@n313893254 thanks for the merge in fix.

I'm noticing that on:
- Rancher v2.7-head
- RKE2 v1.24.9+rke2r2
- Harvester Version: v1.1.2-rc1
That I can see the banner displayed correctly.
![Screenshot from 2023-02-16 12-34-06](https://user-images.githubusercontent.com/5370752/219486171-87583d24-fe37-466a-9971-20ca8436d5c2.png)

On:
- Rancher v2.7-head
- RKE2 v1.25.6+rke2r1
- Harvester Version: v1.1.2-rc1

That when creating the PV and looking at the PVC in Longhorn it looks like it's putting it to `defaultdisk` instead of the desired Storage Class.
The PV -> "Customize" -> "Assign to Storage Class", seems to have it set automatically at "harvester"
 
![Screenshot from 2023-02-16 12-58-45](https://user-images.githubusercontent.com/5370752/219486686-a7471d72-887b-4df6-868e-bc7c37f7d0cb.png)
![Screenshot from 2023-02-16 12-57-46](https://user-images.githubusercontent.com/5370752/219486687-f292eddc-3c9d-46d1-8498-54d90b38c34b.png)
![Screenshot from 2023-02-16 12-56-47](https://user-images.githubusercontent.com/5370752/219486690-72391124-0b45-4f31-9fc9-df424c069d8c.png)
![Screenshot from 2023-02-16 12-56-17](https://user-images.githubusercontent.com/5370752/219486691-54cbf9f7-2bb5-4269-a898-c7e971da9d85.png)
![Screenshot from 2023-02-16 12-55-30](https://user-images.githubusercontent.com/5370752/219486694-1d6dc61a-d5d1-44ee-8d5b-8c47a512cbd7.png)
![Screenshot from 2023-02-16 13-08-07](https://user-images.githubusercontent.com/5370752/219487372-ad54c560-17dd-4b76-87e7-5d37e602fe14.png)
![Screenshot from 2023-02-16 15-33-51](https://user-images.githubusercontent.com/5370752/219511793-23da13b0-3d01-4470-88d9-86ca60a7c962.png)


[supportbundle_7e5e4f7c-03fc-48fd-bc19-31eb63dd96db_2023-02-16T21-00-26Z.zip](https://github.com/harvester/harvester/files/10760851/supportbundle_7e5e4f7c-03fc-48fd-bc19-31eb63dd96db_2023-02-16T21-00-26Z.zip)

And set `ui-index` to https://releases.rancher.com/harvester-ui/dashboard/latest/index.html and `ui-plugin-index` to https://releases.rancher.com/harvester-ui/plugin/harvester-latest/harvester-latest.umd.min.js

**Comment by irishgordo** [comment]
On:
- Rancher v2.6-head
- RKE2 v1.24.9+rke2r2
- Harvester Version: v1.1.2-rc1

That I can see the banner displayed correctly :smile: 
![Screenshot from 2023-02-16 13-59-47](https://user-images.githubusercontent.com/5370752/219498399-6c571a80-bbad-4348-9742-f83a0542c58b.png)

**Comment by irishgordo** [comment]
@n313893254 is there a way to upgrade to harvester-csi-driver 0.1.1500?

I was noticing:
```
RUN CHART_VERSION="0.1.1500" CHART_FILE=/charts/harvester-csi-driver.yaml CHART_BOOTSTRAP=true /charts/build-chart.sh
```
in:
https://github.com/rancher/rke2/pull/3784/files

But wasn't seeing a version higher off `dev-v2.6`:
https://github.com/rancher/charts/blob/dev-v2.6/index.yaml#L1215-#L1226

So I was wondering how we'd update CSI Driver to 0.1.1500 on Rancher v2.6-head to be able to get the latest version to avoid seeing the "Please upgrade your Harvester CSI driver version to use this feature." message - with: Kubernetes Version: v1.24.9 +rke2r2?

**Comment by n313893254** [comment]
Hi, @irishgordo, Thanks for your testing!

We do not recommend that users upgrade their CSI manually, we recommend that they upgrade their RKE2 version. If a user does this, we will tell them to go through "Edit as YAML" or Kubectl as a workaround.

**Comment by irishgordo** [comment]
Hi @n313893254 , I'm glad I could help.

Oh okay, in that case it's likely then the user would just upgrade to the RKE2 version.

**Comment by irishgordo** [comment]
**Validation Failed**

@n313893254 - I'm having some issues validating that an RKE2 cluster can provision with the `harvester-csi-driver` of "0.1.1500".

I was curious what version of KDM we needed to use with the respective Rancher releases.

For [Rancher v2.6.11-rc8](https://hub.docker.com/layers/rancher/rancher/v2.6.11-rc8/images/sha256-21dbd218c8a9cfb5ce78e65e3516b0eb929f15d02e1225fb50ba85c7ca083255?context=explore), Image Layer 11 - mentions `ARG CATTLE_KDM_BRANCH=dev-v2.6`
Which:
```
curl https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/dev-v2.6/data/data.json | jq '.rke2.releases[] | select( .charts["harvester-csi-driver"]["version"] == "0.1.1500" ) | {"rke2-releases-rke2-version": .version, "rke2-releases-rke2-provided-harvester-csi-driver": .charts["harvester-csi-driver"]}'
```
Yields no results, searching for a harvester-csi-driver that meets the requirements of "0.1.1500".

For [Rancher v2.7.1](https://hub.docker.com/layers/rancher/rancher/v2.7.1/images/sha256-45331efbc8a3f9a6a5f3040b2e0e9020581395ed666ad8b1da4c66ca02fba59f?context=explore), Image Layer 11 - mentions `ARG CATTLE_KDM_BRANCH=release-v2.7`
Which:
```
curl https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/release-v2.7/data/data.json | jq '.rke2.releases[] | select( .charts["harvester-csi-driver"]["version"] == "0.1.1500" ) | {"rke2-releases-rke2-version": .version, "rke2-releases-rke2-provided-harvester-csi-driver": .charts["harvester-csi-driver"]}'
```
Yields no results, searching for a `harvester-csi-driver` that meets the requirements of "0.1.1500".

When we look at other KDM branches/forks.

I've noticed that kinarashah's fork at the branch of rel seems to yield the appropriate results of harvester-csi-driver at "0.1.1500", via:
```
curl https://raw.githubusercontent.com/kinarashah/kontainer-driver-metadata/rel/data/data.json | jq '.rke2.releases[] | select( .charts["harvester-csi-driver"]["version"] == "0.1.1500" ) | {"rke2-releases-rke2-version": .version, "rke2-releases-rke2-provided-harvester-csi-driver": .charts["harvester-csi-driver"]}' 
```
Also the instance of the `january_2023_rke2` kdm via:
```
curl https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/january_2023_rke2/data/data.json | jq '.rke2.releases[] | select( .charts["harvester-csi-driver"]["version"] == "0.1.1500" ) | {"rke2-releases-rke2-version": .version, "rke2-releases-rke2-provided-harvester-csi-driver": .charts["harvester-csi-driver"]}'
```
does also (both of them) yield the appropriate results.

Yielding both:
```
╭─mike at suse-workstation-team-harvester in ~
╰─○ curl https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/january_2023_rke2/data/data.json | jq '.rke2.releases[] | select( .charts["harvester-csi-driver"]["version"] == "0.1.1500" ) | {"rke2-releases-rke2-version": .version, "rke2-releases-rke2-provided-harvester-csi-driver": .charts["harvester-csi-driver"]}'

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 4815k  100 4815k    0     0  8939k      0 --:--:-- --:--:-- --:--:-- 8951k
{
  "rke2-releases-rke2-version": "v1.23.16+rke2r1",
  "rke2-releases-rke2-provided-harvester-csi-driver": {
    "repo": "rancher-rke2-charts",
    "version": "0.1.1500"
  }
}
{
  "rke2-releases-rke2-version": "v1.24.10+rke2r1",
  "rke2-releases-rke2-provided-harvester-csi-driver": {
    "repo": "rancher-rke2-charts",
    "version": "0.1.1500"
  }
}
{
  "rke2-releases-rke2-version": "v1.25.6+rke2r1",
  "rke2-releases-rke2-provided-harvester-csi-driver": {
    "repo": "rancher-rke2-charts",
    "version": "0.1.1500"
  }
}
╭─mike at suse-workstation-team-harvester in ~
╰─○ curl https://raw.githubusercontent.com/kinarashah/kontainer-driver-metadata/rel/data/data.json | jq '.rke2.releases[] | select( .charts["harvester-csi-driver"]["version"] == "0.1.1500" ) | {"rke2-releases-rke2-version": .version, "rke2-releases-rke2-provided-harvester-csi-driver": .charts["harvester-csi-driver"]}'
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 5311k  100 5311k    0     0  25.4M      0 --:--:-- --:--:-- --:--:-- 25.4M
{
  "rke2-releases-rke2-version": "v1.23.16+rke2r1",
  "rke2-releases-rke2-provided-harvester-csi-driver": {
    "repo": "rancher-rke2-charts",
    "version": "0.1.1500"
  }
}
{
  "rke2-releases-rke2-version": "v1.24.10+rke2r1",
  "rke2-releases-rke2-provided-harvester-csi-driver": {
    "repo": "rancher-rke2-charts",
    "version": "0.1.1500"
  }
}
{
  "rke2-releases-rke2-version": "v1.25.6+rke2r1",
  "rke2-releases-rke2-provided-harvester-csi-driver": {
    "repo": "rancher-rke2-charts",
    "version": "0.1.1500"
  }
}
```

With Rancher v2.6.11-rc8 when updating the KDM to [january_2023_rke2's data.json](https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/january_2023_rke2/data/data.json)
Causes:
```
2023/02/18 01:02:54 [INFO] driverMetadata: refreshing data from upstream https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/january_2023_rke2/data/data.json
2023/02/18 01:02:54 [INFO] Retrieve data.json from local path /var/lib/rancher-data/driver-metadata/data.json
2023/02/18 01:02:54 [ERROR] API error response 500 for POST /v3/kontainerdrivers?action=refresh. Cause: failed to create or update driverMetadata: driverMetadata: unable to find default k8s version in current k8s v1.25.x [v1.18.20-rancher1-3 v1.19.16-rancher2-1 v1.20.15-rancher2-2 v1.21.14-rancher1-1 v1.22.17-rancher1-1 v1.23.15-rancher1-1 v1.24.9-rancher1-1]
```
(500 errors in Rancher Logs)

With Rancher v2.6.11-rc8 when updating the KDM to [https://raw.githubusercontent.com/kinarashah/kontainer-driver-metadata/rel/data/data.json](https://raw.githubusercontent.com/kinarashah/kontainer-driver-metadata/rel/data/data.json)
That does seem to work with:
```
2023/02/18 01:04:14 [INFO] Refreshing driverMetadata in 10 minutes
2023/02/18 01:04:14 [INFO] getReleaseConfig: reloading config for k3s
2023/02/18 01:04:14 [INFO] getReleaseConfig: reloading config for rke2
2023/02/18 01:04:15 [INFO] driverMetadata: refreshing data from upstream https://raw.githubusercontent.com/kinarashah/kontainer-driver-metadata/rel/data/data.json
2023/02/18 01:04:15 [INFO] Retrieve data.json from local path /var/lib/rancher-data/driver-metadata/data.json
2023/02/18 01:04:16 [INFO] Loaded configuration from https://raw.githubusercontent.com/kinarashah/kontainer-driver-metadata/rel/data/data.json in [0x705ad00]
2023/02/18 01:04:16 [INFO] Loaded configuration from https://raw.githubusercontent.com/kinarashah/kontainer-driver-metadata/rel/data/data.json in [0x705ad00]
2023/02/18 01:04:29 [INFO] driverMetadata: refreshing data from upstream https://raw.githubusercontent.com/kinarashah/kontainer-driver-metadata/rel/data/data.json
2023/02/18 01:04:29 [INFO] Retrieve data.json from local path /var/lib/rancher-data/driver-metadata/data.json
2023/02/18 01:04:29 [INFO] getReleaseConfig: reloading config for k3s
2023/02/18 01:04:29 [INFO] getReleaseConfig: reloading config for rke2
2023/02/18 01:04:30 [INFO] Loaded configuration from https://raw.githubusercontent.com/kinarashah/kontainer-driver-metadata/rel/data/data.json in [0x705ad00]
2023/02/18 01:04:30 [INFO] Loaded configuration from https://raw.githubusercontent.com/kinarashah/kontainer-driver-metadata/rel/data/data.json in [0x705ad00]
```

With Rancher v2.7.1 the:
[https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/january_2023_rke2/data/data.json](https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/january_2023_rke2/data/data.json)
Does seem to work well with the logs showing:
```
2023/02/18 01:19:45 [INFO] Refreshing driverMetadata in 10 minutes
2023/02/18 01:19:45 [INFO] getReleaseConfig: reloading config for k3s
2023/02/18 01:19:45 [INFO] getReleaseConfig: reloading config for rke2
2023/02/18 01:19:45 [INFO] driverMetadata: refreshing data from upstream https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/january_2023_rke2/data/data.json
2023/02/18 01:19:45 [INFO] Retrieve data.json from local path /var/lib/rancher-data/driver-metadata/data.json
2023/02/18 01:19:46 [INFO] Loaded configuration from https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/january_2023_rke2/data/data.json in [0x6d10098]
2023/02/18 01:19:46 [INFO] Loaded configuration from https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/january_2023_rke2/data/data.json in [0x6d10098]
```

But both in Rancher v2.6.11-rc8 & Rancher v2.7.1 - I am unable to provision an RKE2 cluster that would have the `harvester-csi-driver` at version "0.1.1500" - leveraging the corresponding RKE2 versions that would have said driver:
```
curl https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/january_2023_rke2/data/data.json | jq '.rke2.releases[] | select( .charts["harvester-csi-driver"]["version"] == "0.1.1500" ) | {"rke2-releases-rke2-version": .version, "rke2-releases-rke2-provided-harvester-csi-driver": .charts["harvester-csi-driver"]}'

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 4815k  100 4815k    0     0  8939k      0 --:--:-- --:--:-- --:--:-- 8951k
{
  "rke2-releases-rke2-version": "v1.23.16+rke2r1",
  "rke2-releases-rke2-provided-harvester-csi-driver": {
    "repo": "rancher-rke2-charts",
    "version": "0.1.1500"
  }
}
{
  "rke2-releases-rke2-version": "v1.24.10+rke2r1",
  "rke2-releases-rke2-provided-harvester-csi-driver": {
    "repo": "rancher-rke2-charts",
    "version": "0.1.1500"
  }
}
{
  "rke2-releases-rke2-version": "v1.25.6+rke2r1",
  "rke2-releases-rke2-provided-harvester-csi-driver": {
    "repo": "rancher-rke2-charts",
    "version": "0.1.1500"
  }
}
```
As the Harvester Cloud Provider selection is entirely greyed out on both.

**Screenshots attached:**
![Screenshot from 2023-02-17 17-12-39](https://user-images.githubusercontent.com/5370752/219824861-c1a98760-985a-4a30-9b9a-50899e4b4353.png)
![Screenshot from 2023-02-17 17-13-07](https://user-images.githubusercontent.com/5370752/219824863-d681f77d-cc81-4021-8ecd-7387e4eeb08c.png)
![Screenshot from 2023-02-17 17-22-37](https://user-images.githubusercontent.com/5370752/219824865-a55eb209-01d9-43a0-a89b-97b3ebc69768.png)
![Screenshot from 2023-02-17 17-23-09](https://user-images.githubusercontent.com/5370752/219824867-032c6e58-a097-4043-87d9-897877ffa8b3.png)

**Comment by n313893254** [comment]
@irishgordo Same as an issue [3389](https://github.com/harvester/harvester/issues/3389). The Harvester cloud provider can not select [issue](https://github.com/harvester/harvester/issues/3404) is caused by the version setting check: `/v1/harvester/harvesterhci.io.settings/harvester-csi-ccm-versions`. 

~The workaround is to update the setting default value or value to `{"harvester-cloud-provider":">=0.0.1","harvester-csi-provider":">=0.0.1"}`~

The workaround is to change the Harvester server-version setting to v1.1.2.

![image](https://user-images.githubusercontent.com/18737885/220052565-8546c0e1-127b-426b-adf0-6742ebe6b98b.png)

**Comment by irishgordo** [comment]
**validation failure**

@n313893254 - that's a good catch :smile: 

I went ahead and implemented the change via the "View in API" -> "Edit" -> "Send Request".

I'm still noticing that with v2.7.1 Rancher that both options that have a "harvester-csi-driver" at version "0.1.1500" , the Harvester cloud provider is greyed out.

I'll check v2.6.11-rc8 to see if there's any difference as well.

Is there another workaround that needs to be implemented?
I also removed the cluster from Rancher, then re-added, to see if that had any effect and it didn't seem to.

![Screenshot from 2023-02-21 11-58-44](https://user-images.githubusercontent.com/5370752/220448802-f9e21954-ef4d-4154-ba75-2c1bbd9deb74.png)
![Screenshot from 2023-02-21 11-58-18](https://user-images.githubusercontent.com/5370752/220448803-b4659495-cded-43aa-86fc-32c8d438a713.png)

**Comment by irishgordo** [comment]
**validation with v2.6.8-rc8, question / seems-to-have passed**

@n313893254 
For Rancher v2.6.8-rc8
I was able to provision an RKE2 v1.23.16+rke2r1 single node cluster
On Harvester.

I noticed that the `harvester-csi-driver` seems to not be at `0.1.500` however though:
![Screenshot from 2023-02-21 13-21-25](https://user-images.githubusercontent.com/5370752/220462087-252741a1-4917-43c6-879a-4d2868a2f2c2.png)
As it shows `v0.1.4` but should we be expecting `v0.1.5`?

I did notice though that a PVC built on the cluster and the respecting PV seem to be respecting the storage class created.
![Screenshot from 2023-02-21 13-28-07](https://user-images.githubusercontent.com/5370752/220462500-7fddeaf0-c332-4411-abe6-8f4de6432a5a.png)
![Screenshot from 2023-02-21 13-24-50](https://user-images.githubusercontent.com/5370752/220462502-122c2dd2-2f46-4196-9fb2-32e40555c30b.png)
![Screenshot from 2023-02-21 13-24-21](https://user-images.githubusercontent.com/5370752/220462504-5e73b672-c652-478c-9b60-c9765b7f9259.png)

**Comment by n313893254** [comment]
@irishgordo 
> The Harvester cloud provider is greyed out in Rancher v2.7.1

As the PR [8080](https://github.com/rancher/dashboard/pull/8080) was merged after v2.7.1 was released. For Rancher v2.7.1 and other versions less than v2.7.2-rc2 / v2.6.11-rc4. The workaround is to update the setting default value or value to `{"harvester-cloud-provider":">=0.0.1","harvester-csi-provider":">=0.0.1"}`

**Comment by irishgordo** [comment]
@n313893254 thanks for the response back :smile: 

Testing on v2.7.2-rc2 Rancher ( using RKE Metadata config of: https://raw.githubusercontent.com/rancher/kontainer-driver-metadata/january_2023_rke2/data/data.json & Kubernetes Version: v1.24.10 ) noticing that we can create a storage class on the RKE2 cluster that points to the Storage Class created in Harvester - then create a PVC, that builds a PV, and the corresponding PVC is appropriately located on the extra-disks - related to the storage class on Harvester :smile: 

![Screenshot from 2023-02-22 12-52-27](https://user-images.githubusercontent.com/5370752/220757425-82fd9971-31f8-4f34-952d-2fa55ebe0d27.png)
![Screenshot from 2023-02-22 12-41-11](https://user-images.githubusercontent.com/5370752/220757430-1782cd0d-2829-41e4-86dd-df2b28eb4064.png)

I'll go ahead and close this out :+1:
